nextcord/cog.py

```diff
@@ -0,0 +1,338 @@
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    ClassVar,
+    Dict,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+)
+
+from .enums import ApplicationCommandOptionType, ApplicationCommandType
+from .utils import MISSING
+
+if TYPE_CHECKING:
+    from .application_command import BaseApplicationCommand
+    from .interactions import Interaction
+
+__all__ = (
+    "CogMeta",
+    "Cog",
+    "ClientCog",
+)
+
+FuncT = TypeVar("FuncT", bound=Callable[..., Any])
+
+
+def _cog_special_method(func: FuncT) -> FuncT:
+    func.__cog_special_method__ = None
+    return func
+
+
+class CogMeta(type):
+    """A metaclass for defining a cog.
+
+    Note that you should probably not use this directly. It is exposed
+    purely for documentation purposes along with making custom metaclasses to intermix
+    with other metaclasses such as the :class:`abc.ABCMeta` metaclass.
+
+    For example, to create an abstract cog mixin class, the following would be done.
+
+    .. code-block:: python3
+
+        import abc
+
+        class CogABCMeta(nextcord.CogMeta, abc.ABCMeta):
+            pass
+
+        class SomeMixin(metaclass=abc.ABCMeta):
+            pass
+
+        class SomeCogMixin(SomeMixin, nextcord.Cog, metaclass=CogABCMeta):
+            pass
+
+    .. note::
+
+        When passing an attribute of a metaclass that is documented below, note
+        that you must pass it as a keyword-only argument to the class creation
+        like the following example:
+
+        .. code-block:: python3
+
+            class MyCog(nextcord.Cog, name="My Cog"):
+                pass
+
+    Attributes
+    ----------
+    name: :class:`str`
+        The cog name. By default, it is the name of the class with no modification.
+    description: :class:`str`
+        The cog description. By default, it is the cleaned docstring of the class.
+    """
+
+    __cog_name__: str
+
+    def __new__(cls: Type[CogMeta], *args: Any, **kwargs: Any) -> CogMeta:
+        name, bases, attrs = args
+        attrs["__cog_name__"] = kwargs.pop("name", name)
+
+        description = kwargs.pop("description", None)
+        if description is None:
+            description = inspect.cleandoc(attrs.get("__doc__", ""))
+        attrs["__cog_description__"] = description
+
+        return super().__new__(cls, name, bases, attrs, **kwargs)
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args)
+
+    @classmethod
+    def qualified_name(cls) -> str:
+        return cls.__cog_name__
+
+
+class Cog(metaclass=CogMeta):
+    """The base class that all cogs must inherit from.
+
+    A cog is a collection of commands, listeners, and optional state to
+    help group commands together. More information on them can be found on
+    the :ref:`cogs` page.
+
+    When inheriting from this class, the options shown in :class:`nextcord.CogMeta`
+    are equally valid here.
+
+    There is an alias for this called :class:`ClientCog`.
```

Comment:
I feel like there's no need to state this given its deprecated and will be removed?